braille-printer-app/brf-printer-app.c

```diff
@@ -0,0 +1,1071 @@
+
+  brf_cups_device_data_t *device_data = NULL;
+  // brf_job_data_t * job_data;
+  const char *informat;
+  const char *filename;                  // Input filename
+  int fd;                                // Input file descriptor
+  brf_spooling_conversion_t *conversion; // Spooling conversion to use for pre-filtering
+  cups_array_t *spooling_conversions;
+  cf_filter_filter_in_chain_t *chain_filter, // Filter from PPD file
+      *print;
+  cf_filter_external_t *filter_data_ext;
+  brf_print_filter_function_data_t *print_params;
+  cf_filter_data_t *filter_data;
+  cups_array_t *chain;
+  int nullfd; // File descriptor for /dev/null
+  char paramstr[1024];
+  char buf[1024];
+
+  bool ret = false;    // Return value
+  int num_options = 0; // Number of PPD print options
+
+  pappl_pr_options_t *job_options = papplJobCreatePrintOptions(job, INT_MAX, 0);
+
+  cups_option_t *options = NULL;
+  cf_filter_external_t *ext_filter_params;
+
+  pappl_pr_driver_data_t driver_data;
+  pappl_printer_t *printer = papplJobGetPrinter(job);
+  const char *device_uri = papplPrinterGetDeviceURI(printer);
+
+  ipp_t *driver_attrs = papplPrinterGetDriverAttributes(printer);
+
+  paramstr[sizeof(paramstr) - 1] = 0;
+
+  // Define an array of option names
+  const char *option_names[] = {"PageSize","mirror","fitplot",
+                                  "SendFF", "SendSUB",
+                                 "LibLouis", "LibLouis2", "LibLouis3", "LibLouis4",
+                                "TextDotDistance", "TextDots", "LineSpacing", "TopMargin", "BottomMargin",
+                                "LeftMargin", "RightMargin", "BraillePageNumber", "PrintPageNumber",
+                                "PageSeparator", "PageSeparatorNumber", "ContinuePages", "GraphicDotDistance",
+                                "Rotate", "Edge", "Negate", "EdgeFactor", "CannyRadius", "CannySigma",
+                                "CannyLower", "CannyUpper", "page-left", "page-right", "page-top", "page-bottom"};
+
+  // Loop through each option and process them
+  for (size_t i = 0; i < sizeof(option_names) / sizeof(option_names[0]); i++)
+  {
+    const char *option_name = option_names[i];
+    ipp_attribute_t *attribute = papplJobGetAttribute(job, option_name);
+
+   // If attribute is not found, look for the default attribute
+    if (attribute == NULL) {
+        snprintf(buf, sizeof(buf), "%s-default", option_name);
+        attribute = ippFindAttribute(driver_attrs, buf, IPP_TAG_ZERO);
+        if (attribute != NULL) {
+            printf("Default attribute for %s: %s\n", option_name, ippGetName(attribute));
+        }
+    }
+
+    // Try to retrieve supported attribute as well
+    snprintf(buf, sizeof(buf), "%s-supported", option_name);
+    ipp_attribute_t *attr_support = ippFindAttribute(driver_attrs, buf, IPP_TAG_ZERO);
+    if (attr_support != NULL) {
+        printf("Supported attribute for %s: %s\n", option_name, ippGetName(attr_support));
+    }
```

Comment:
We will not be using it so it's not useful to get this `%s-supported` attribute.